Add an installer script resource type and public HTTP endpoint

lxea opened the pull request:
This adds an installer script resource type, this will be used by the teleport aws discovery feature (RFD #12410).

Part of this adds a HTTP endpoint that doesnt require authentication so as it can be fetched by nodes that are not yet in the cluster, to install teleport.

Commit:
Add an installer script resource type and HTTP endpoint

marcoandredinis commented:
> Where is the best place to add tests for a new resources/their create/delete testing?
> 
> Also I tried to be quite thorough with adding the resource in the right places as necessary though I'm not sure I missed somewhere or not. `tctl create -f installer.yml`, `tctl get installer` and `tctl delete installer` all work.

That's probably too much effort but adding this resource to Terraform Provider :sunglasses: 
Would be a good test

marcoandredinis commented on api/types/installer.go:
```diff
@@ -0,0 +1,122 @@
+func (c *InstallerV1) setStaticFields() {
+	c.Kind = KindInstaller
+	c.Version = V1
+	c.Metadata.Name = MetaNameClusterAuditConfig
```
```suggestion
	c.Metadata.Name = MetaNameClusterInstallerScript
```

marcoandredinis commented on api/types/types.proto:
```diff
@@ -3538,6 +3538,25 @@ message Participant {
+// InstallerSpecV1 is the specification for an Installer
+message InstallerSpecV1 {
+    // Script represents the contents of a installer shell script
+    string Script = 1 [ (gogoproto.jsontag) = "script" ];
```
This `script` is a bot opaque on what it really is 
I think we could at least add some metadata to it (either the scripting language or the binary name)

AON, the RFD has an `expire` field which is not present here

marcoandredinis commented:
Do we support multiple installer scripts? or is it a singleton/config like resource?

I can see this being used in other places (not saying it will :joy:)

marcoandredinis commented on lib/cache/collections.go:
```diff
@@ -87,6 +87,11 @@ func setupCollections(c *Cache, watches []types.WatchKind) (map[resourceKind]col
 			}
 			collections[resourceKind] = &sessionRecordingConfig{watch: watch, Cache: c}
+		case types.KindInstaller:
+			if c.ClusterConfig == nil {
+				return nil, trace.BadParameter("missing parameter ClusterConfig")
```
Typo here

marcoandredinis commented on lib/cache/collections.go:
```diff
@@ -2160,6 +2165,74 @@ func (c *sessionRecordingConfig) watchKind() types.WatchKind {
+		// either zero or one instance exists, so we either erase or
+		// update, but not both.
+		if noConfig {
+			if err := c.erase(ctx); err != nil {
+				return trace.Wrap(err)
+			}
+			return nil
```
```suggestion
			return trace.Wrap(c.erase(ctx))
```

marcoandredinis commented on lib/cache/collections.go:
```diff
@@ -2160,6 +2165,74 @@ func (c *sessionRecordingConfig) watchKind() types.WatchKind {
+			return nil
+		}
+
+		if err := c.clusterConfigCache.SetInstaller(ctx, resource); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
```
```suggestion
		return trace.Wrap(c.clusterConfigCache.SetInstaller(ctx, resource))
```

marcoandredinis commented:
I would like to test this, can I do it now? Or would you like to rebase and fix the failing tests (they are probably flaky ones) first?

marcoandredinis commented:
@lxea 
Sorry, I only got to this today
It seems the conflicting file (`api/client/proto/authservice.pb.go`) is not allowing for a complete build

marcoandredinis reviewed (approved):
LGTM

Just one question, is this message expected?

If we try to get the installer without anything created we get
```shell
root@marco:/workspace# tctl get installer
ERROR: key /cluster_configuration/installer_script is not found
```

Couldn't find other resources that return this error message
Most singletons resources have some pre-defined instantiation
Other resources return nothing when they have 0 elements

If this is common pattern :+1: 
Otherwise, I think it would be worth to add a different message
Something like: you haven't configured the installer, please do so like this: ... (or link to docs)

---
`tctl get installer` :heavy_check_mark: 

I think the command to delete the resource should be `tctl rm` instead of `tctl delete` like you have in the comment in this PR:

```shell
root@marco:/workspace# tctl delete installer
tctl: error: expected command but got "delete"
root@marco:/workspace# tctl rm installer
installer has been deleted
root@marco:/workspace# tctl get installer
ERROR: key /cluster_configuration/installer_script is not found

root@marco:/workspace#
```

r0mant commented on lib/web/apiserver.go:
```diff
@@ -300,6 +300,10 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 
+	// Unauthenticated access to retrieving the script used to install
+	// Teleport
+	h.GET("/webapi/install-teleport.sh", httplib.MakeHandler(h.installer))
```
Can we tuck this under "scripts" or something, and also do we need `.sh` suffix? `/webapi/scripts/installer`.

r0mant commented on tool/tctl/common/resource_command.go:
```diff
@@ -782,6 +794,12 @@ func (rc *ResourceCommand) Delete(ctx context.Context, client auth.ClientI) (err
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Printf("%s has been deleted\n", types.KindInstaller)
```
What does "delete" do? Doesn't it restore the default installer script? If so, the message should probably say that.

r0mant commented on tool/tctl/common/resource_command.go:
```diff
@@ -1249,6 +1267,15 @@ func (rc *ResourceCommand) getCollection(ctx context.Context, client auth.Client
 		return &tokenCollection{tokens: []types.ProvisionToken{token}}, nil
+	case types.KindInstaller:
+		if rc.ref.Name != "" {
+			return nil, trace.BadParameter("only simple `tctl get %v` can be used", types.KindInstaller)
```
```suggestion
			return nil, trace.BadParameter("installer is a singleton resource, use `tctl get %v` to fetch it", types.KindInstaller)
```

r0mant replied:
Seems like this comment got resolved without actually applying the suggestion.

r0mant reviewed (changes requested):
I think RBAC (auth_with_roles) needs a bit more work, otherwise looks alright.

r0mant reviewed (commented):
@lxea A few more action items on this PR:

- [x] Limit proxy's permissions for installer script to read-only (unless it needs r/w).
- [x] Adjust the default script behavior to make sure there is always a script present and can't be deleted. Maybe we don't need an actual resource to always be present, the API can just return a default resource if there's no installer resource present in backend.
- [x] Verify trusted clusters scenario and make sure there are no caching related errors when an older cluster is connected to the cluster running version built off of this branch.
- [x] Add test coverage, there are no tests on this PR currently. We want to test basic CRUD on the installer script (including the scenario with the default script I mentioned above) and RBAC.

Commit:
Add default installer scripts

rosstimothy reviewed (commented):
Was `lib/web/scripts/installer.sh.tmpl` meant to be included?

r0mant commented on api/types/constants.go:
```diff
@@ -182,6 +182,10 @@ const (
 	MetaNameClusterNetworkingConfig = "cluster-networking-config"
 
+	// MetaNameClusterInstallerScript is the exact name of the
+	// singleton resource holding cluster audit configuration
```
Comment needs updating probably ("cluster audit configuration").

r0mant commented on api/types/installer.go:
```diff
@@ -0,0 +1,122 @@
+			Script: script,
+		},
+	}
+	return installer
```
Do we need to call `CheckAndSetDefaults()` on installer before returning? Otherwise we're returning technically an incorrect resource.

r0mant commented on api/types/installers/installer.sh.tmpl:
```diff
@@ -0,0 +1,38 @@
+  fi
+
+  distro_id="$(awk -F= '$1 == "ID" { print tolower($2) }' /etc/os-release | xargs echo)"
+  if [ "$distro_id" = "debian" ] || [ "$distro_id" = "ubuntu" ]; then
+          sudo curl https://deb.releases.teleport.dev/teleport-pubkey.asc \
+                   -o /usr/share/keyrings/teleport-archive-keyring.asc
+          echo "deb [signed-by=/usr/share/keyrings/teleport-archive-keyring.asc] https://deb.releases.teleport.dev/ stable main" \
+                  | sudo tee /etc/apt/sources.list.d/teleport.list > /dev/null
+
+          sudo apt-get update
+          sudo apt-get install -y teleport jq
+  elif [ "$distro_id" = "amzn" ] || [ "$distro_id" = "rhel" ]; then
+          sudo yum-config-manager --add-repo 'https://rpm.releases.teleport.dev/teleport.repo'
+          sudo yum install -y teleport jq
+  fi
```
Couple notes:

1. This should use the new APT repositories (https://goteleport.com/docs/installation/#linux). New RPM repos are also coming out this week, so we'll need to update yum instructions also.
2. I don't see where we specify the Teleport version that is being installed. We want to install the same version that auth server is running as we discussed before via an install parameter, right?

r0mant commented on api/types/installers/installer.sh.tmpl:
```diff
@@ -0,0 +1,38 @@
+                   --auth-server="{{ .AuthServer }}" \
+                   --join-method=iam \
+                   --token="$1" \
+                   --node-name="${ACCOUNT_ID}-${INSTANCE_ID}" | sudo tee "/etc/teleport.yaml"
```
Why do we need to set the node name?

r0mant commented on api/types/installers/installers.go:
```diff
@@ -0,0 +1,32 @@
+var DefaultInstaller = types.NewInstallerV1(defaultInstallScript)
+
+type InstallerTemplate struct {
+	AuthServer string
```
Yeah, we will definitely be adding documentation for this.

r0mant commented on lib/auth/permissions.go:
```diff
@@ -322,6 +322,7 @@ func (a *authorizer) authorizeRemoteBuiltinRole(r RemoteBuiltinRole) (*Context,
 					types.NewRule(types.KindSessionRecordingConfig, services.RO()),
 					types.NewRule(types.KindClusterAuthPreference, services.RO()),
 					types.NewRule(types.KindKubeService, services.RO()),
+					types.NewRule(types.KindProxy, services.RO()),
```
Why is this permission on `KindProxy` needed?

r0mant commented on lib/services/presets.go:
```diff
@@ -68,6 +68,7 @@ func NewPresetEditorRole() types.Role {
 					types.NewRule(types.KindTrustedCluster, RW()),
 					types.NewRule(types.KindRemoteCluster, RW()),
 					types.NewRule(types.KindToken, RW()),
+					types.NewRule(types.KindInstaller, RW()),
```
Yes, I mixed it up with permissions.go. This should be ok.

r0mant commented on tool/tctl/common/resource_command.go:
```diff
@@ -782,6 +794,12 @@ func (rc *ResourceCommand) Delete(ctx context.Context, client auth.ClientI) (err
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		fmt.Printf("%s has been reset to a default value\n", types.KindInstaller)
```
Where is the logic that actually returns the default installer value if it's not set? I either didn't see it or missed it. I only saw that logic that sets installer if it's not set on the init which is not the correct place for it because resource can still be deleted and won't be recreated until after restart.

r0mant commented on lib/auth/init.go:
```diff
@@ -274,6 +275,10 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	if err := initSetDefaultInstaller(ctx, asrv); err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Why do we need to set anything in the backend? Why not just do this:

- When installer is being retrieved, try to fetch it from the backend.
 - If found, return it.
 - If not found, return a default installer resource (just build the resource in memory).

Then we won't need to save any singletons in the backend and won't need to worry about re-saving the default installer resource when user deletes their own. Does that make sense?

rosstimothy reviewed (commented):
Are the webassets changes intended to be included here? 

Also for future consideration, it is helpful to your reviewers if you don't squash commits during the review. It makes figuring out what has changed during the life of a PR rather challenging, especially for PRs that are open for a long period of time.

r0mant commented on lib/service/service.go:
```diff
@@ -3382,6 +3383,9 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 				ClusterFeatures:  process.getClusterFeatures(),
 				ProxySettings:    proxySettings,
+				Installer: installers.InstallerTemplate{
+					AuthServer: process.proxyPublicAddr().Addr,
```
I think it's fine to default to connect through proxy since we want this to work out of the box with Cloud. We can just document how to connect directly to auth for whoever wants it.

r0mant commented on api/types/installers/installer.sh.tmpl:
```diff
@@ -0,0 +1,41 @@
+  # token is read as a parameter from the AWS ssm script run and
+  # passed as the first argument to the script
+  /usr/local/bin/teleport node configure \
+						  --auth-server="{{ .PublicProxyAddr }}" \
+						  --join-method=iam \
+						  --token="$1" \
+						  --labels="teleport.dev/instance-id=${INSTANCE_ID},teleport.dev/account-id=${ACCOUNT_ID}" \
```
Nit: Indentation seems weird here (too many whitespaces).

r0mant commented on api/types/installers/installer.sh.tmpl:
```diff
@@ -0,0 +1,41 @@
+						  --join-method=iam \
+						  --token="$1" \
+						  --labels="teleport.dev/instance-id=${INSTANCE_ID},teleport.dev/account-id=${ACCOUNT_ID}" \
+						  --node-name="${ACCOUNT_ID}-${INSTANCE_ID}" | sudo tee "/etc/teleport.yaml"
```
Why do we need to set the node name explicitly? What if we omit this flag here?

r0mant replied:
Also, I think that the "configure" command accepts `-o=file` flag that will write it to /etc/teleport.yaml so you don't need to pipe to tee.

r0mant replied:
Hmm I see, I think with auto-discovery we don't expect the config to exist so overwriting is probably fine. Let's keep your approach for now.

r0mant commented on api/types/installers/installer.sh.tmpl:
```diff
@@ -0,0 +1,41 @@
+	  sudo yum-config-manager --add-repo \
+		   $(rpm --eval "https://yum.releases.teleport.dev/$ID/$VERSION_ID/Teleport/%{_arch}/stable/{{ .MajorVersion }}/teleport.repo")
+	  sudo yum install -y teleport jq
+  fi
```
What if it's any other distro? We should exit with some error.

r0mant commented on api/types/installers/installers.go:
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2015-2021 Gravitational, Inc.
```
```suggestion
Copyright 2022 Gravitational, Inc.
```

r0mant commented on api/types/installers/installers.go:
```diff
@@ -0,0 +1,35 @@
+// by teleport
+var DefaultInstaller, _ = types.NewInstallerV1(defaultInstallScript)
+
+type Template struct {
```
Nit: Missing godoc.

r0mant commented on api/types/installers/installers.go:
```diff
@@ -0,0 +1,35 @@
+
+// DefaultInstaller represents a the default installer script provided
+// by teleport
+var DefaultInstaller, _ = types.NewInstallerV1(defaultInstallScript)
```
Instead of ignoring the error, I would make a `MustNewInstallerV1` method that would call `NewInstallerV1` and panic on error.

r0mant commented on lib/web/apiserver.go:
```diff
@@ -1379,6 +1388,33 @@ func (h *Handler) oidcCallback(w http.ResponseWriter, r *http.Request, p httprou
+	httplib.SetScriptHeaders(w.Header())
+	installer, err := h.auth.proxyClient.GetInstaller(r.Context())
+	if err != nil {
+		if trace.IsNotFound(err) {
+			installer = installers.DefaultInstaller
+		} else {
+			return nil, trace.Wrap(err)
+		}
```
I don't think this fallback logic belongs in the web handler. Otherwise e.g. tctl get installer will return "not found" if you don't have the installer. This fallback should be implemented in `GetInstaller` so if you call that API method and there's not custom installer, a default object is returned.

r0mant commented on lib/web/apiserver.go:
```diff
@@ -1379,6 +1388,33 @@ func (h *Handler) oidcCallback(w http.ResponseWriter, r *http.Request, p httprou
+	}
+	ping, err := h.auth.Ping(r.Context())
+	if err != nil {
+		return nil, err
```
```suggestion
		return nil, trace.Wrap(err)
```

r0mant reviewed (commented):
Hopefully the final batch of comments.

Commit:
Resolve comments

- run `shfmt` over the installer script
- add MustNewInstaller
- move default installer logic to GetInstaller
- only include PublicProxyAddr in server.Config

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3357,6 +3358,37 @@ func (a *ServerWithRoles) GetAuthPreference(ctx context.Context) (types.AuthPref
+		if trace.IsNotFound(err) {
+			return installers.DefaultInstaller, nil
+		}
+		return nil, err
```
```suggestion
		return nil, trace.Wrap(err)
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3357,6 +3358,37 @@ func (a *ServerWithRoles) GetAuthPreference(ctx context.Context) (types.AuthPref
+	}
+	installer, err := a.authServer.GetInstaller(ctx)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return installers.DefaultInstaller, nil
+		}
```
I'd probably push it into actual auth server's GetInstaller method, not sure auth_with_roles which is an authz layer is the appropriate place for this fallback.

r0mant commented on api/types/installer.go:
```diff
@@ -0,0 +1,133 @@
+	return installer, nil
+}
+
+func MustNewInstallerV1(script string) *InstallerV1 {
```
```suggestion
// MustNewInstallerV1 creates a new installer resource from the provided script.
//
// Panics in case of any error when creating the resource.
func MustNewInstallerV1(script string) *InstallerV1 {
```

r0mant reviewed (approved):
Looks good to me now with a couple remaining nitpicks. Thank you for putting up with many rounds of code review!

@tigrato @rosstimothy Do you guys have any other feedback on this one?

Commit:
Resolve comments

- move from auth_with_roles -> auth/grpcserver
- document `MustNewInstallerV1`

Commit:
Allow remote proxy to read KindInstaller
